### DISCUSSION TITLE
Restoring <Editable /> value after validation in onSubmit

### DISCUSSION THREAD

**Discussion by TomasB** [chakra-ui/chakra-ui#3176] [Q&A]
How would one do error checking in `<Editable />`  control?
Say, if I were to implement editing in place, and sending a request to update server side as soon as onSubmit happens, how would I be able to restore the value to the initial value?
Or, similar applies to validation - how can I prevent submission, and restore value if value isn't acceptable?

**Comment by Vichp** [comment]
Excellent question, I would also like to know that!

**Comment by dimi-colossal** [comment]
I'm having the same issue - there isn't a way to validate and avoid submission in the first place. There's also no way to get the value of the `<Editable>` and apply some validation to the nested `<EditableInput>` (I'm using a custom `<Input>`). Also, you cannot forward the `value` to your custom `EditableControls` to disable the submit button.

**Reply by dimi-colossal** [reply]
@segunadebayo maybe you can help us out

**Comment by nicoalee** [comment]
hey @TomasB I was facing the same issue and this is the solution I came up with. I know your answer is 3 years old, but maybe this will help someone else:

```
function MyEditableControl({ textToEdit, onChange }: { textToEdit: string; onChange: (text: string) => void }) {

    ...

    return (
        <Editable value={textToEdit} onChange={handleChange}>
            <Box display="flex" justifyContent="space-between" w="100%" alignItems="center">
                <Box flexGrow={1} mr="4">
                    <EditablePreview />
                    <Input as={EditableInput} />
                </Box>
                <Box>
                    <EditableControls textToEdit={value} />
                </Box>
            </Box>
        </Editable>
    )
}

function EditableControls({ textToEdit }: { textToEdit: string }) {
    const { isEditing, getSubmitButtonProps, getCancelButtonProps, getEditButtonProps } = useEditableControls();
    const submitButtonProps = getSubmitButtonProps();

    const isValid = (): boolean => {
        // TODO: do validation check on textToEdit and return true/false
    }

    return (
        <>
            {isEditing ? (
                <ButtonGroup size="sm">
                    <IconButton
                        aria-label="save"
                        icon={<CheckIcon />}
                        {...submitButtonProps}
                        onClick={(e) => {
                           if (!isValid()) {
                              // TODO: show error message and error styling
                              return;
                           }
                            if (submitButtonProps.onClick) submitButtonProps.onClick(e);
                        }}
                    ></IconButton>
                    <IconButton
                        colorScheme="red"
                        aria-label="cancel"
                        icon={<CloseIcon />}
                        {...cancelButtonProps}
                    ></IconButton>
                </ButtonGroup>
            ) : (
                <Box>
                    <IconButton
                        colorScheme="secondary"
                        aria-label="edit"
                        size="sm"
                        icon={<EditIcon size="20" />}
                        {...editButtonProps}
                    ></IconButton>
                </Box>
            )}
        </>
    )

}
```

You would then pass the string you are editing to the MyEditableControl component from some parent, and update it when you receive a call to the onChange callback. If you click save and some server call fails, you can keep track of the original value and revert it.

**Comment by vls9** [comment]
One way to reset the `Editable` would be to change its `key` prop with `useState`, thus forcing the `Editable` to re-render:

`setEditableKey((prevKey) => prevKey + 1);`

I know this is kinda hackish, but it appears to be the easiest solution without redesigning the entire input form.

@nicoalee @TomasB @dimi-colossal
